core/src/hir/lowering.rs

```diff
@@ -778,32 +808,374 @@ fn lower_return_type(
                         Ok(None)
                     }
                 }
-                ty => lower_out_type(ty, parent_lifetimes, lookup_id, in_path, env, errors)
+                ty => lower_out_type(ty, Ok(&mut return_ltl), lookup_id, in_path, env, errors)
                     .map(|ty| Some(ReturnType::OutType(ty))),
             };
 
-            let err_ty = lower_out_type(err_ty, parent_lifetimes, lookup_id, in_path, env, errors);
+            let err_ty =
+                lower_out_type(err_ty, Ok(&mut return_ltl), lookup_id, in_path, env, errors);
 
             match (ok_ty, err_ty) {
-                (Ok(ok_ty), Ok(err_ty)) => Ok(ReturnFallability::Fallible(ok_ty, err_ty)),
+                (Ok(ok_ty), Ok(err_ty)) => Ok((
+                    ReturnFallability::Fallible(ok_ty, err_ty),
+                    return_ltl.finish(),
+                )),
                 _ => Err(()),
             }
-        }
-        ast::TypeName::Unit => {
+        }),
+        ast::TypeName::Unit => return_ltl.map(|return_ltl| {
             if takes_writeable {
-                Ok(ReturnFallability::Infallible(Some(ReturnType::Writeable)))
+                (
+                    ReturnFallability::Infallible(Some(ReturnType::Writeable)),
+                    return_ltl.finish(),
+                )
             } else {
-                Ok(ReturnFallability::Infallible(None))
+                (ReturnFallability::Infallible(None), return_ltl.finish())
+            }
+        }),
+        _ => return_ltl.and_then(|mut return_ltl| {
+            lower_out_type(
+                return_type,
+                Ok(&mut return_ltl),
+                lookup_id,
+                in_path,
+                env,
+                errors,
+            )
+            .map(|ty| {
+                (
+                    ReturnFallability::Infallible(Some(ReturnType::OutType(ty))),
+                    return_ltl.finish(),
+                )
+            })
+        }),
+    }
+}
+
+pub mod elision {
```

Comment:
nit: probably ought to a separate file (doesn't need to be a submodule IMO)

Reply:
All of this is still specifically related to lowering, so I felt like this was appropriate. I could definitely move it to a separate file though.

I agree that this seriously needs docs, but I'd rather write them once I resolve the corner case I mentioned. The PR for that fix (and these docs) should be pretty quick once this PR is merged since I already know the fix and didn't want to bloat this PR too much.